thriftbp/admin.go

```diff
@@ -13,5 +13,5 @@ import (
 //
 // This function blocks, so it should be run as its own goroutine.
 func ServeAdmin() {
-	log.Warnw("thriftbp: admin serving exited", "err", admin.Serve())
+	log.Panicw("thriftbp: admin serving exited", "err", admin.Serve())
```

Comment:
I don't believe this is the best way to do it.

There are historical reasons that we don't make this log level higher: the http server's `Serve` function [always returns non-nil error eventually](https://pkg.go.dev/net/http#Server.Serve). Currently we do not actually ever shutdown the admin http server so it won't happen, but if we ever change the `internal/admin`'s api to shut it down during graceful shutdown, this panic will cause the graceful shutdown process to panic and interrupt other cleanups.

The thing you want to guard against is that the 6060 port is unavailable so `admin.Serve` fails early (vs. normal shutdown). To achieve that, we should actually check the error returned by `admin.Serve` and only panic if it's not [`http.ErrServerClosed`](https://pkg.go.dev/net/http#ErrServerClosed). Something like this:

```go
if err := admin.Serve(); errors.Is(err, http.ErrServerClosed) {
  log.Infow(...)
} else {
  log.Panicw(...)
}
```

Also you want to do the same in `httpbp.ServeAdmin` as well.

(another historical reason of not make the log level higher is that admin port used to only serve pprof, which is really not critical enough to bring the whole pod down. now with prometheus I think it's justified)

Reply:
+1 on that